### PR TITLE
pf4/wp5: Migrate dashboard to PF4

### PR DESCRIPTION
**Pull request by sjd78**
:exclamation: _this PR has been replaced by #70_  :exclamation:
----

## Migrate Dashboard to PF4, remove PF3, align CSS

Bug-Url: https://bugzilla.redhat.com/show_bug.cgi?id=2021123
Note: Built on (and currently includes) PR #65.
Note: This PR removes all use of PF3, Bootstrap and jQuery from the project enabling the upgrade to webpack 5 and EL9 builds to follow.

### Screen shots

Fake data dashboard:
screenshot-engine-dev ovirt_8080-2022 07 06-10_13_40

Fake data dashboard loading:
screenshot-engine-dev ovirt_8080-2022 07 06-10_14_58

HeatMap tooltip:
screenshot-engine-dev ovirt_8080-2022 07 06-10_16_01

Utilization popups:
screenshot-engine-dev ovirt_8080-2022 07 06-10_21_59

### Dashboard changes

Project dependencies changes:
  - remove patternfly (v3),
  - remove patternfly-react (v3),
  - and resolutions that are no longer needed

PF refactoring on dashboard:
  - Dashboard imports now are consistent with only PF4 imports, PF3 imports have been removed

  - The initial data loading spinner and the loading error message are converted to PF4 and render properly in the horizontal and vertical center of the iframe page

  - Refactor `RefreshDataControls` from pf3 to pf4-react

  - Upgrade `HeatMap` to a functional component and use PF4 Tooltips:
    - Convert `HeatMap` to a functional component
    - Drop use of bootstrap and jquery for Tooltips
    - Drop use of D3 for rendering the `HeatMap`'s SVG
    - Adopt use of PF4 Tooltips attached to every SVG `rect` generated

  - Handle layout changes needed after switching to the pf4 reset css

The base CSS used by plugin, modals (loaded from the plugin iframe), and the dashboard page are now aligned.  They all use the same base `plugin-pf4-overrides.css` to make the pages look closer to the PF3 styles of webadmin.

### Underlying changes
  - Upgrade PF4 packages to 2022.07 release
    Closest previous release: 2020.13 (2020-10-06)
    New Release: 2022.07 (2022-06-03)
    React Notes: https://www.patternfly.org/v4/developer-resources/release-notes#2022.07-release-notes-2022-06-02
    HTML Notes: https://www.patternfly.org/v4/developer-resources/release-notes/html#2022.07-release-notes-2022-06-03

  - Upgrade `css-element-queries` used in `HeightMatching` from 0.4.0 to 1.2.3

  - Remove c3, jquery and bootstrap, HeatMap will not need them

  - Drop some unnecessary jest polyfills

  - Remove logger.js, rely on fancy-console babel plugin

  - DashboardDataProvider use fake-data sleep to show the loading spinner

  - To aid in visual testing, always render the `MonitoringPortalLink` on dashboard when `config.useFakeData` is enabled.  If graphana is not configured on the dev engine, the url is rendered with a simple javascript console log.

### CODE REVIEW

File: src/dashboard/MonitoringPortalLink.js [hunk 1]
```diff
@@ -1,11 +1,26 @@
...
 import { msg } from '_/intl-messages'
+import config from '_/plugin-config'
+
+import { Button } from '@patternfly/react-core'
 
 const MonitoringPortalLink = ({ url }) => {
-  return url && (
-    <a className='monitoring-portal-link' target='_top' href={url}>{msg.dashboardLinkMonitoringPortal()}</a>
-  )
+  const onClick = url
```
Comment (rszwajko) [inline src/dashboard/MonitoringPortalLink.js:9]: we could simplify the flow here i.e.
```
if (!url && !config.useFakeData) {
   return null
}
const onClick = () => {
  url &&  window.open(...) 
  config.useFakeData && console.info(...)
}
return <Button ...>
```

File: src/dashboard/RefreshDataControl.js [hunk 2]
```diff
@@ -2,21 +2,19 @@ import React from 'react'
...
-        >
-          <i className='fa fa-refresh' />
-        </button>
-        {/* refresh configuration drop down menu would go here */}
-      </div>
+    <Tooltip content={msg.dashboardRefreshButtonTooltip()} position='right-start' distance={5}>
+      <Button
+        className='refreshButton'
+        variant='plain'
+        onClick={() => onRefresh()}
```
Comment (rszwajko) [inline src/dashboard/RefreshDataControl.js:14]: `onClick={onRefresh}` ?

File: src/components/patternfly/HeatMap.js [hunk 3]
```diff
@@ -1,115 +1,117 @@
...
+    height: 100,
+  },
+  onBlockClick = () => {},
+}) => {
+  const containerRef = useRef()
+  const [{ blockSize, numberOfRows }, setRenderParams] = useState({})
+  const [allBlocksActive, setAllBlocksActive] = useState(true)
+  const [mouseOverBlock, setMouseOverBlock] = useState(-1)
+
+  useEffect(() => {
```
Comment (rszwajko) [inline src/components/patternfly/HeatMap.js:57]: why we need `useEffect` hook? if this is only for performance then perhaps we could use `useMemo`?

**Review by rszwajko — COMMENTED** [review]
Looks good!

File: src/components/patternfly/HeatMap.js (hunk 3 above)
Comment (sjd78) [reply src/components/patternfly/HeatMap.js:57]: The effect uses the `containerRef` and needs to run AFTER the initial render is done and the DOM is all updated.  The ref won't exist until then.

From https://reactjs.org/docs/hooks-effect.html:

> What does useEffect do? By using this Hook, you tell React that your component needs to do something after render. React will remember the function you passed (we’ll refer to it as our “effect”), and call it later after performing the DOM updates. In this effect, we set the document title, but we could also perform data fetching or call some other imperative API.

The important lines that need the ref:
```jsx
  useEffect(() => {
    const containerWidth = containerRef.current.clientWidth
    const containerHeight = containerRef.current.clientHeight
```

Comment (rszwajko) [reply src/components/patternfly/HeatMap.js:57]: Ok. Tha makes sense. However combining element size detection with block calculations makes this algorithm look complex. How about decoupling them? i.e. similar as [useElementSize](https://usehooks-ts.com/react-hook/use-element-size) hook. We could also consider re-using the hook (or similar).

Comment (rszwajko) [reply src/components/patternfly/HeatMap.js:57]: @sjd78 I had some visual problems after resizing the window. Not sure if this is due to this part but the values are calculated after resize (the hook depends only on `data` and  `maxBlockSize`).

Comment (sjd78) [reply src/components/patternfly/HeatMap.js:57]: Nice hook, I'll use that.

**Comment by sjd78** [comment]
:exclamation: Closing this PR since it has been superseded by #70.

**Pull request closed without merging**